Implement policy finding candidates to be slashed

linked0 opened the pull request:
This creates the policy manager class for slashing and defined the base criteria where a validator will be slashed as the first step of issue 1076. As a result of the whiteboarding session, we start with 10k on the missing pre-image, effective immediately. 

This PR does not contain how to impose penalty on the misbehaving validator yet. I will implement the feature imposing 10K BOA on the validator in other PR.

Part of #1076

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -0,0 +1,159 @@
+
+    public Amount getPenaltyAmount()
+    {
+        return this.penalty_amount;
```
I'm not sure this belongs here. Will this class be in charge of slashing? It seems not, because it has a public method `getSlashCandidates`, so it will be another class which does the actual slashing right? This one is just about selecting the candidates.

But then this class doesn't need to store the penalty amount, the other class will be in charge of that.

linked0 replied:
The description of the class is confusing. I will update it. Actually this class is in charge of managing slashing policy which determine the condition where the slashing must be done.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -519,6 +519,20 @@ public class EnrollmentManager
+
+    ***************************************************************************/
+
+    PreImageInfo[] getAllPreimages () @trusted nothrow
```
`public`. And please use `@safe` and make the function in `ValidatorSet` also `@safe`.

linked0 replied:
My mistake. I will update it.

AndrejMitrovic commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -414,6 +414,39 @@ public class ValidatorSet
+
+    ***************************************************************************/
+
+    PreImageInfo[] getAllPreimages () @trusted nothrow
```
`public` and `@safe`

linked0 replied:
Ditto.

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -0,0 +1,159 @@
+        if the last revealed pre-image is behind, based on the current height.
+
+        Params:
+            block_height = current block height
```
```suggestion
            height = current block height
```

linked0 replied:
Will update it.

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -0,0 +1,159 @@
+            if (preimage.distance < curr_distance)
+            {
+                UTXO utxo_value;
+                finder(preimage.enroll_key, utxo_value);
```
UTXOFinder returns a boolean, please use it here instead of comparing with `UTXO.init`.

linked0 replied:
Ditto

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -0,0 +1,159 @@
+    public this (EnrollmentManager enroll_man)
+    {
+        this.enroll_man = enroll_man;
+        this.penalty_amount = Amount(100_000_000_000);
```
This ultimately belongs in the `ConsensusParams`. However, since this is WIP you may keep it hardcoded for now.

linked0 replied:
Currently, the `SlashPolicy` class manages the amount but it can be moved into `ConsensusParams` as you commented. It will be done on other PR.

linked0 committed:
Add `getAllPreimages` function to `EnrollmentManager`

This function will be used to check the pre-images hav been revealed timely.
After getting pre-images of all the validators, we prepare the distance of
pre-images with required distance based on current height.

linked0 committed:
Implement policy finding candidates to be slashed

This creates the policy manager class for slashing and defined the base
criteria where a validator will be slashed as first step of the issue 1076.

AndrejMitrovic reviewed (changes requested):
Just a few issues.

linked0 commented:
This PR is not needed anymore.
